src/main/java/org/opensearch/plugin/insights/rules/transport/top_queries/TransportTopQueriesAction.java

```diff
@@ -68,27 +73,122 @@ public TransportTopQueriesAction(
         this.queryInsightsService = queryInsightsService;
     }
 
+    protected ActionListener<TopQueriesResponse> createInMemoryDataCollectionListener(
+        TopQueriesRequest request,
+        ActionListener<TopQueriesResponse> finalListener
+    ) {
+        return new ActionListener<TopQueriesResponse>() {
+            @Override
+            public void onResponse(TopQueriesResponse inMemoryQueriesResponse) {
+                handleInMemoryDataResponse(request, inMemoryQueriesResponse, finalListener);
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                finalListener.onFailure(e);
+            }
+        };
+    }
+
+    protected void handleInMemoryDataResponse(
+        TopQueriesRequest request,
+        TopQueriesResponse inMemoryQueriesResponse,
+        ActionListener<TopQueriesResponse> finalListener
+    ) {
+        List<TopQueries> inMemoryTopQueries = inMemoryQueriesResponse.getNodes();
+        List<FailedNodeException> inMemoryDataFailures = inMemoryQueriesResponse.failures();
+        String from = request.getFrom();
+        String to = request.getTo();
+        if (from != null && to != null) {
+            fetchHistoricalData(request, inMemoryTopQueries, inMemoryDataFailures, finalListener);
+        } else {
+            finalListener.onResponse(inMemoryQueriesResponse);
+        }
+    }
+
+    protected void fetchHistoricalData(
+        TopQueriesRequest request,
+        List<TopQueries> inMemoryTopQueries,
+        List<FailedNodeException> inMemoryDataFailures,
+        ActionListener<TopQueriesResponse> finalListener
+    ) {
+        String from = request.getFrom();
+        String to = request.getTo();
+        String id = request.getId();
+        Boolean verbose = request.getVerbose();
+        final ThreadContext.StoredContext storedContext = threadPool.getThreadContext().stashContext();
```

Comment:
I think better way to stash the context is using `try ()` block it will automatically restore it once try block is finished executing because `StoredContext` is `AutoCloseable`.